File: node/classes/FontStreamer.js
// FontStreamer
// Takes a font file and streams bytes/words sequentially
const _ = require('lodash');
const BitArray = require('node-bitarray');

class FontStreamer {
    constructor(rawFontFile, readPosition) {
        this.rawFontFile = rawFontFile;
        // strip the first 32 bytes off to make the pointer locations accurate
        this.fontFile = rawFontFile.slice(32);
        // font data starts at position 78
        this.readPosition = readPosition || 0;
    }

    setReadPosition(readPosition) {
        this.readPosition = readPosition;
    }

    readNextByte() {
        const byte = this.fontFile.readUInt8(this.readPosition);
        this.readPosition += 1;
        return byte;
    }

    readNextWord() {
        const word = this.fontFile.readUInt16BE(this.readPosition);
        this.readPosition += 2;
        return word;
    }

    readNextPointer() {
        const ptr = this.fontFile.readUInt32BE(this.readPosition);
        this.readPosition += 4;
        return ptr;
    }

    getBytesAt(start, length) {
        return this.fontFile.slice(start, length ? start + length : undefined);
    }

    getBitArray = (pointer, modulo, ySize) => {
        const fontBitmapData = this.getBytesAt(pointer, modulo * ySize);
        const fontBitArray = BitArray.fromBuffer(fontBitmapData).toJSON();
        return  _.chunk(fontBitArray, modulo * 8);
    }
    
}

module.exports = FontStreamer

File: python/classes/FontStreamer.py
# FontStreamer
# Takes a font file and streams bytes/words sequentially

from utils import chunks

class FontStreamer:
    def __init__(self, rawFontFile, readPosition):
        self.rawFontFile = rawFontFile
        # strip the first 32 bytes off to make the pointer locations accurate
        self.fontFile = rawFontFile[32:]
        self.readPosition = readPosition if readPosition else 0


    def setReadPosition(self, readPosition):
        self.readPosition = readPosition


    def readNextByte(self):
        byte = int.from_bytes(self.fontFile[self.readPosition:self.readPosition + 1], byteorder='big', signed=False)
        self.readPosition += 1
        return byte


    def readNextWord(self):
        word = int.from_bytes(self.fontFile[self.readPosition:self.readPosition + 2], byteorder='big', signed=False)
        self.readPosition += 2
        return word


    def readNextPointer(self):
        ptr = int.from_bytes(self.fontFile[self.readPosition:self.readPosition + 4], byteorder='big', signed=False)
        self.readPosition += 4
        return ptr


    def getBytesAt(self, start, length=None):
        if length:
            return self.fontFile[start:start + length]
        
        return self.fontFile[start:]

    def getBitArray(self, pointer, modulo, ySize):
        fontBitmapData = self.getBytesAt(pointer, modulo * ySize)
        # From https://stackoverflow.com/questions/43787031/python-byte-array-to-bit-array
        fontBitArray = ''.join(format(byte, '08b') for byte in fontBitmapData)
        chunkedArray = list(chunks(fontBitArray, modulo * 8))
        return [list(row) for row in chunkedArray]
